src/compiler/moduleNameResolver.ts

```diff
@@ -2420,7 +2421,16 @@ function loadModuleFromSelfNameReference(extensions: Extensions, moduleName: str
     // in order to be consistent with (non-self) library-name loading in
     // `loadModuleFromNearestNodeModulesDirectoryWorker`, which uses two passes in order
     // to prioritize `@types` packages higher up the directory tree over untyped
-    // implementation packages.
+    // implementation packages. See the selfNameModuleAugmentation.ts test for why this
+    // matters.
+    //
+    // However, there's an exception. If the user has `allowJs` and `declaration, we need
+    // to ensure that self-name imports of their own package can resolve back to their
+    // input JS files via `tryLoadInputFileForPath` at a higher priority than their output
+    // declaration files, so we need to do a single pass with all extensions for that case.
+    if (getAllowJSCompilerOption(state.compilerOptions) && !pathContainsNodeModules(directory)) {

```

Comment:
I think with this we need to add affectsModuleResolution to "checkJs" in commandline options declaration.

Reply:
Ooh, yikes. I can do that for now, but maybe it would be better to update `getCompilerOptionValue` (called by `compilerOptionsAffect**` functions) to use all the computed compiler option getter functions.

Reply:
`checkJs` doesn’t even have `affectsSemanticDiagnostics`. Does it need it?

Reply:
it does. #53403 never made it in i guess.

Reply:
I’ll leave it off of this one and then we should revisit #53403